# Task 2.4 - Optimizing Hyperparameters - CNN Model

## Table of Contents

#### 1. Importing Libraries and Data
- Cleaned weather observations from 2.2, pleasant weather (predictions)
#### 2. Data Wrangling
#### 3. Reshaping for Modeling
#### 4. Data Split
#### 5. Hyperparameter Optimization
-  Bayesian optimization
#### 6. Running CNN with Optimized Search Parameters
#### 7. Creating Confusion Matrix

## 1. Importing Libraries and Data

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import os
import operator
import time
import matplotlib.pyplot as plt
from sklearn.preprocessing import LabelEncoder
import tensorflow as tf
from numpy import unique
from numpy import reshape
from keras.models import Sequential
from sklearn.model_selection import cross_val_score
from keras.layers import Conv1D, Conv2D, Dense, Dropout, BatchNormalization, Flatten, MaxPooling1D
from tensorflow.keras.utils import to_categorical
from keras.optimizers import Adam, SGD, RMSprop, Adadelta, Adagrad, Adamax, Nadam, Ftrl
from keras.callbacks import EarlyStopping, ModelCheckpoint
from scikeras.wrappers import KerasClassifier
from math import floor
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.metrics import make_scorer, accuracy_score
from bayes_opt import BayesianOptimization
from sklearn.model_selection import StratifiedKFold
from keras.layers import LeakyReLU
LeakyReLU = LeakyReLU(alpha=0.1)

import warnings
warnings.filterwarnings('ignore')

pd.set_option("display.max_columns", None)

2024-09-13 11:54:35.935672: I tensorflow/core/platform/cpu_feature_guard.cc:210] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.
/Users/giadairene/anaconda3/lib/python3.11/site-packages/keras/src/layers/activations/leaky_relu.py:41: UserWarning: Argument `alpha` is deprecated. Use `negative_slope` instead.
  warnings.warn(


In [2]:
# Set a path

path = r'/Users/giadairene/Documents/CareerFoundry Data Analytics/Machine Learning with Python/ClimateWins'

In [3]:
# Import cleaned weather observations data

X = pd.read_csv(os.path.join(path, '02 Data', 'Prepared', 'X_cleaned_date.csv'), index_col = False)

In [4]:
# Import predictions data

answers = pd.read_csv(os.path.join(path, '02 Data', 'Original', 'Dataset-Answers-Weather_Prediction_Pleasant_Weather.csv'), index_col = False)

In [5]:
X.head()

DATE  MONTH  BASEL_cloud_cover  BASEL_humidity  BASEL_pressure  \
0  19600101      1                  7            0.85           1.018   
1  19600102      1                  6            0.84           1.018   
2  19600103      1                  8            0.90           1.018   
3  19600104      1                  3            0.92           1.018   
4  19600105      1                  6            0.95           1.018   

   BASEL_global_radiation  BASEL_precipitation  BASEL_sunshine  \
0                    0.32                 0.09             0.7   
1                    0.36                 1.05             1.1   
2                    0.18                 0.30             0.0   
3                    0.58                 0.00             4.1   
4                    0.65                 0.14             5.4   

   BASEL_temp_mean  BASEL_temp_min  BASEL_temp_max  BELGRADE_cloud_cover  \
0              6.5             0.8            10.9                     1   
1              6.1             3.3            10.1                     6   
2              8.5             5.1             9.9                     6   
3              6.3             3.8            10.6                     8   
4              3.0            -0.7             6.0                     8   

   BELGRADE_humidity  BELGRADE_pressure  BELGRADE_global_radiation  \
0               0.81             1.0195                       0.88   
1               0.84             1.0172                       0.25   
2               0.77             1.0179                       0.67   
3               0.93             1.0268                       0.25   
4               0.99             1.0286                       0.25   

   BELGRADE_precipitation  BELGRADE_sunshine  BELGRADE_temp_mean  \
0                    0.00                7.0                 3.7   
1                    0.00                0.0                 2.9   
2                    0.00                3.5                 3.1   
3                    0.00                0.0                 2.0   
4                    0.06                0.0                 2.0   

   BELGRADE_temp_min  BELGRADE_temp_max  BUDAPEST_cloud_cover  \
0               -0.9                7.9                     4   
1                2.2                4.4                     4   
2               -0.5                6.4                     4   
3               -2.0                3.0                     4   
4                0.7                2.8                     4   

   BUDAPEST_humidity  BUDAPEST_pressure  BUDAPEST_global_radiation  \
0               0.67              1.017                       0.44   
1               0.67              1.017                       0.18   
2               0.67              1.017                       0.30   
3               0.67              1.017                       0.19   
4               0.67              1.017                       0.19   

   BUDAPEST_precipitation  BUDAPEST_sunshine  BUDAPEST_temp_mean  \
0                    0.01                2.3                 2.4   
1                    0.31                0.0                 2.3   
2                    0.00                0.6                 2.7   
3                    0.00                0.0                 2.0   
4                    0.00                0.0                 2.5   

   BUDAPEST_temp_min  BUDAPEST_temp_max  DEBILT_cloud_cover  DEBILT_humidity  \
0               -0.4                5.1                   7             0.85   
1                1.4                3.1                   8             0.90   
2                1.7                5.3                   6             0.92   
3                0.4                4.4                   8             0.95   
4                1.1                5.3                   6             0.90   

   DEBILT_pressure  DEBILT_global_radiation  DEBILT_precipitation  \
0           1.0032                     0.07                  0.25   
1           1.0056                     0.14 

In [6]:
X.shape

(22950, 137)

In [7]:
answers.head()

DATE  BASEL_pleasant_weather  BELGRADE_pleasant_weather  \
0  19600101                       0                          0   
1  19600102                       0                          0   
2  19600103                       0                          0   
3  19600104                       0                          0   
4  19600105                       0                          0   

   BUDAPEST_pleasant_weather  DEBILT_pleasant_weather  \
0                          0                        0   
1                          0                        0   
2                          0                        0   
3                          0                        0   
4                          0                        0   

   DUSSELDORF_pleasant_weather  HEATHROW_pleasant_weather  \
0                            0                          0   
1                            0                          0   
2                            0                          0   
3                            0                          0   
4                            0                          0   

   KASSEL_pleasant_weather  LJUBLJANA_pleasant_weather  \
0                        0                           0   
1                        0                           0   
2                        0                           0   
3                        0                           0   
4                        0                           0   

   MAASTRICHT_pleasant_weather  MADRID_pleasant_weather  \
0                            0                        0   
1                            0                        0   
2                            0                        0   
3                            0                        0   
4                            0                        0   

   MUNCHENB_pleasant_weather  OSLO_pleasant_weather  \
0                          0                      0   
1                          0                      0   
2                          0                      0   
3                          0                      0   
4                          0                      0   

   SONNBLICK_pleasant_weather  STOCKHOLM_pleasant_weather  \
0                           0                           0   
1                           0                           0   
2                           0                           0   
3                           0                           0   
4                           0                           0   

   VALENTIA_pleasant_weather  
0                          0  
1                          0  
2                          0  
3                          0  
4                          0

In [8]:
answers.shape

(22950, 16)

## 2. Data Wrangling

In [9]:
# Drop unnecessary columns

X.drop(['DATE', 'MONTH'], axis=1, inplace=True)

In [10]:
X.shape # observations dataset has the correct shape

(22950, 135)

In [11]:
answers.drop(columns = 'DATE', inplace = True)

In [12]:
answers.shape # predictions dataset has the correct shape

(22950, 15)

## 3. Reshaping for Modeling
- Your X shape will need to be in the form (22950, 15, 9)
- Your y shape will need to be in the form (22950,)
- You can use the argmax function to transform your y set

In [13]:
# Turn X and answers from a df to arrays

X = np.array(X)
y = np.array(answers)

In [14]:
X = X.reshape(-1,15,9)

In [15]:
# Verify shape

X.shape

(22950, 15, 9)

In [16]:
# Use argmax to transform y

y =  np.argmax(y, axis = 1)
y

array([0, 0, 0, ..., 0, 0, 0])

In [17]:
y.shape

(22950,)

In [18]:
# Check y layout

from sklearn.utils.multiclass import type_of_target
type_of_target(y)

'multiclass'

#### The Bayesian optimization function only accepts y data in “multiclass” and “binary” layouts but not in “multilabel-indicator.”

## 4. Data Split 

In [19]:
# Split data into train and test sets

X_train, X_test, y_train, y_test = train_test_split(X,y,random_state = 42)

In [20]:
print(X_train.shape, y_train.shape)
print(X_test.shape, y_test.shape)

(17212, 15, 9) (17212,)
(5738, 15, 9) (5738,)


## 5. Bayesian Hyperparameter Optimization

In [21]:
timesteps = len(X_train[0])
input_dim = len(X_train[0][0])
n_classes = 15 # Number of weather stations
# Make scorer accuracy
score_acc = make_scorer(accuracy_score)

In [22]:
# Create function

def bay_area(neurons, activation, kernel, optimizer, learning_rate, batch_size, epochs,
              layers1, layers2, normalization, dropout, dropout_rate): 
    optimizerL = ['SGD', 'Adam', 'RMSprop', 'Adadelta', 'Adagrad', 'Adamax', 'Nadam', 'Ftrl','SGD']
    #optimizerD= {'Adam':Adam(lr=learning_rate), 'SGD':SGD(lr=learning_rate),
                 #'RMSprop':RMSprop(lr=learning_rate), 'Adadelta':Adadelta(lr=learning_rate),
                 #'Adagrad':Adagrad(lr=learning_rate), 'Adamax':Adamax(lr=learning_rate),
                 #'Nadam':Nadam(lr=learning_rate), 'Ftrl':Ftrl(lr=learning_rate)}
    activationL = ['relu', 'sigmoid', 'softplus', 'softsign', 'tanh', 'selu',
                   'elu', 'exponential', LeakyReLU,'relu']
    
    neurons = round(neurons)
    kernel = round(kernel)
    activation = activationL[round(activation)]  #optimizerD[optimizerL[round(optimizer)]]
    optimizer = optimizerL[round(optimizer)]
    batch_size = round(batch_size)
    
    epochs = round(epochs)
    layers1 = round(layers1)
    layers2 = round(layers2)
    
    def cnn_model():
        model = Sequential()
        model.add(Conv1D(neurons, kernel_size=kernel,activation=activation, input_shape=(timesteps, input_dim)))
        #model.add(Conv1D(32, kernel_size=1,activation='relu', input_shape=(timesteps, input_dim)))
        
        if normalization > 0.5:
            model.add(BatchNormalization())
        for i in range(layers1):
            model.add(Dense(neurons, activation=activation)) #(neurons, activation=activation))
        if dropout > 0.5:
            model.add(Dropout(dropout_rate, seed=123))
        for i in range(layers2):
            model.add(Dense(neurons, activation=activation))
        model.add(MaxPooling1D())
        model.add(Flatten())
        model.add(Dense(n_classes, activation='softmax')) #sigmoid softmax
        #model.compile(loss='binary_crossentropy', optimizer=optimizer, metrics=['accuracy']) #categorical_crossentropy
        model.compile(loss='sparse_categorical_crossentropy', optimizer=optimizer, metrics=['accuracy']) #categorical_crossentropy
        return model
    es = EarlyStopping(monitor='accuracy', mode='max', verbose=2, patience=20)
    nn = KerasClassifier(build_fn=cnn_model, epochs=epochs, batch_size=batch_size, verbose=2)
    kfold = StratifiedKFold(n_splits=5, shuffle=True, random_state=123)
    score = cross_val_score(nn, X_train, y_train, scoring=score_acc, cv=kfold, fit_params={'callbacks':[es]}).mean()
    return score

In [23]:
start = time.time()
params ={
    'neurons': (10, 100),
    'kernel': (1, 3),
    'activation':(0, 9), 
    'optimizer':(0,7),
    'learning_rate':(0.01, 1),
    'batch_size': (200, 1000), 
    'epochs':(20, 50),
    'layers1':(1,3),
    'layers2':(1,3),
    'normalization':(0,1),
    'dropout':(0,1),
    'dropout_rate':(0,0.3)
}
# Run Bayesian Optimization
nn_opt = BayesianOptimization(bay_area, params, random_state=42)
nn_opt.maximize(init_points=15, n_iter=4) 
print('Search took %s minutes' % ((time.time() - start)/60))

|   iter    |  target   | activa... | batch_... |  dropout  | dropou... |  epochs   |  kernel   |  layers1  |  layers2  | learni... |  neurons  | normal... | optimizer |
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------
Epoch 1/25
15/15 - 4s - 276ms/step - accuracy: 0.6005 - loss: 2.7005
Epoch 2/25
15/15 - 2s - 115ms/step - accuracy: 0.6440 - loss: 2.7004
Epoch 3/25
15/15 - 2s - 109ms/step - accuracy: 0.6440 - loss: 2.6970
Epoch 4/25
15/15 - 2s - 106ms/step - accuracy: 0.6440 - loss: 2.6941
Epoch 5/25
15/15 - 2s - 102ms/step - accuracy: 0.6440 - loss: 2.6916
Epoch 6/25
15/15 - 1s - 98ms/step - accuracy: 0.6440 - loss: 2.6893
Epoch 7/25
15/15 - 1s - 99ms/step - accuracy: 0.6440 - loss: 2.6872
Epoch 8/25
15/15 - 1s - 97ms/step - accuracy: 0.6440 - loss: 2.6852
Epoch 9/25
15/15 - 3s - 193ms/step - accuracy: 0.6440 - loss: 2.6833
Epoch 10/25
15/15 - 2s - 121ms/step - accu

Epoch 18/25
15/15 - 2s - 118ms/step - accuracy: 0.6440 - loss: 2.6685
Epoch 19/25
15/15 - 2s - 117ms/step - accuracy: 0.6440 - loss: 2.6670
Epoch 20/25
15/15 - 2s - 152ms/step - accuracy: 0.6440 - loss: 2.6655
Epoch 21/25
15/15 - 1s - 97ms/step - accuracy: 0.6440 - loss: 2.6639
Epoch 22/25
15/15 - 2s - 102ms/step - accuracy: 0.6440 - loss: 2.6623
Epoch 23/25
15/15 - 2s - 101ms/step - accuracy: 0.6440 - loss: 2.6607
Epoch 24/25
15/15 - 1s - 100ms/step - accuracy: 0.6440 - loss: 2.6591
Epoch 25/25
15/15 - 2s - 119ms/step - accuracy: 0.6440 - loss: 2.6574
4/4 - 0s - 64ms/step
Epoch 1/25
15/15 - 3s - 224ms/step - accuracy: 0.6005 - loss: 2.7045
Epoch 2/25
15/15 - 1s - 99ms/step - accuracy: 0.6439 - loss: 2.7004
Epoch 3/25
15/15 - 1s - 98ms/step - accuracy: 0.6439 - loss: 2.6970
Epoch 4/25
15/15 - 1s - 98ms/step - accuracy: 0.6439 - loss: 2.6942
Epoch 5/25
15/15 - 3s - 169ms/step - accuracy: 0.6439 - loss: 2.6917
Epoch 6/25
15/15 - 1s - 99ms/step - accuracy: 0.6439 - loss: 2.6894
Epoch 7/25

Epoch 28/29
38/38 - 1s - 16ms/step - accuracy: 0.6439 - loss: nan
Epoch 29/29
38/38 - 0s - 8ms/step - accuracy: 0.6439 - loss: nan
10/10 - 0s - 18ms/step
Epoch 1/29
38/38 - 2s - 56ms/step - accuracy: 0.6440 - loss: nan
Epoch 2/29
38/38 - 0s - 8ms/step - accuracy: 0.6440 - loss: nan
Epoch 3/29
38/38 - 1s - 16ms/step - accuracy: 0.6440 - loss: nan
Epoch 4/29
38/38 - 0s - 8ms/step - accuracy: 0.6440 - loss: nan
Epoch 5/29
38/38 - 0s - 8ms/step - accuracy: 0.6440 - loss: nan
Epoch 6/29
38/38 - 0s - 8ms/step - accuracy: 0.6440 - loss: nan
Epoch 7/29
38/38 - 0s - 8ms/step - accuracy: 0.6440 - loss: nan
Epoch 8/29
38/38 - 0s - 8ms/step - accuracy: 0.6440 - loss: nan
Epoch 9/29
38/38 - 0s - 8ms/step - accuracy: 0.6440 - loss: nan
Epoch 10/29
38/38 - 0s - 8ms/step - accuracy: 0.6440 - loss: nan
Epoch 11/29
38/38 - 0s - 8ms/step - accuracy: 0.6440 - loss: nan
Epoch 12/29
38/38 - 0s - 9ms/step - accuracy: 0.6440 - loss: nan
Epoch 13/29
38/38 - 0s - 8ms/step - accuracy: 0.6440 - loss: nan
Epoch 14

Epoch 19/38
17/17 - 2s - 136ms/step - accuracy: 0.8738 - loss: 0.3652
Epoch 20/38
17/17 - 2s - 144ms/step - accuracy: 0.8813 - loss: 0.3481
Epoch 21/38
17/17 - 2s - 126ms/step - accuracy: 0.8792 - loss: 0.3526
Epoch 22/38
17/17 - 2s - 123ms/step - accuracy: 0.8885 - loss: 0.3261
Epoch 23/38
17/17 - 2s - 121ms/step - accuracy: 0.8908 - loss: 0.3183
Epoch 24/38
17/17 - 2s - 122ms/step - accuracy: 0.8869 - loss: 0.3244
Epoch 25/38
17/17 - 2s - 129ms/step - accuracy: 0.8948 - loss: 0.3058
Epoch 26/38
17/17 - 2s - 127ms/step - accuracy: 0.8974 - loss: 0.2975
Epoch 27/38
17/17 - 2s - 125ms/step - accuracy: 0.8985 - loss: 0.2924
Epoch 28/38
17/17 - 3s - 164ms/step - accuracy: 0.9078 - loss: 0.2724
Epoch 29/38
17/17 - 2s - 143ms/step - accuracy: 0.9061 - loss: 0.2728
Epoch 30/38
17/17 - 2s - 127ms/step - accuracy: 0.9083 - loss: 0.2676
Epoch 31/38
17/17 - 2s - 127ms/step - accuracy: 0.9083 - loss: 0.2685
Epoch 32/38
17/17 - 2s - 127ms/step - accuracy: 0.9109 - loss: 0.2620
Epoch 33/38
17/17 - 

Epoch 22/38
17/17 - 2s - 124ms/step - accuracy: 0.8982 - loss: 0.2999
Epoch 23/38
17/17 - 2s - 127ms/step - accuracy: 0.9036 - loss: 0.2788
Epoch 24/38
17/17 - 3s - 152ms/step - accuracy: 0.9033 - loss: 0.2748
Epoch 25/38
17/17 - 3s - 154ms/step - accuracy: 0.9137 - loss: 0.2581
Epoch 26/38
17/17 - 2s - 131ms/step - accuracy: 0.9133 - loss: 0.2547
Epoch 27/38
17/17 - 2s - 130ms/step - accuracy: 0.9187 - loss: 0.2434
Epoch 28/38
17/17 - 2s - 141ms/step - accuracy: 0.9178 - loss: 0.2448
Epoch 29/38
17/17 - 2s - 142ms/step - accuracy: 0.9217 - loss: 0.2326
Epoch 30/38
17/17 - 2s - 129ms/step - accuracy: 0.9230 - loss: 0.2292
Epoch 31/38
17/17 - 3s - 154ms/step - accuracy: 0.9259 - loss: 0.2187
Epoch 32/38
17/17 - 4s - 217ms/step - accuracy: 0.9261 - loss: 0.2163
Epoch 33/38
17/17 - 5s - 273ms/step - accuracy: 0.9293 - loss: 0.2130
Epoch 34/38
17/17 - 3s - 191ms/step - accuracy: 0.9293 - loss: 0.2082
Epoch 35/38
17/17 - 3s - 171ms/step - accuracy: 0.9325 - loss: 0.2000
Epoch 36/38
17/17 - 

Epoch 1/24
50/50 - 3s - 61ms/step - accuracy: 0.2357 - loss: 2.5250
Epoch 2/24
50/50 - 2s - 34ms/step - accuracy: 0.6306 - loss: 2.0082
Epoch 3/24
50/50 - 2s - 35ms/step - accuracy: 0.6430 - loss: 1.6553
Epoch 4/24
50/50 - 2s - 34ms/step - accuracy: 0.6438 - loss: 1.4469
Epoch 5/24
50/50 - 2s - 33ms/step - accuracy: 0.6439 - loss: 1.3355
Epoch 6/24
50/50 - 2s - 34ms/step - accuracy: 0.6439 - loss: 1.2727
Epoch 7/24
50/50 - 2s - 34ms/step - accuracy: 0.6439 - loss: 1.2334
Epoch 8/24
50/50 - 2s - 38ms/step - accuracy: 0.6439 - loss: 1.2072
Epoch 9/24
50/50 - 2s - 35ms/step - accuracy: 0.6439 - loss: 1.1862
Epoch 10/24
50/50 - 2s - 30ms/step - accuracy: 0.6439 - loss: 1.1700
Epoch 11/24
50/50 - 1s - 29ms/step - accuracy: 0.6439 - loss: 1.1561
Epoch 12/24
50/50 - 2s - 30ms/step - accuracy: 0.6439 - loss: 1.1449
Epoch 13/24
50/50 - 1s - 30ms/step - accuracy: 0.6439 - loss: 1.1348
Epoch 14/24
50/50 - 2s - 34ms/step - accuracy: 0.6439 - loss: 1.1252
Epoch 15/24
50/50 - 2s - 33ms/step - accura

Epoch 43/48
40/40 - 1s - 16ms/step - accuracy: 0.8504 - loss: 0.4178
Epoch 44/48
40/40 - 0s - 11ms/step - accuracy: 0.8439 - loss: 0.4242
Epoch 45/48
40/40 - 0s - 11ms/step - accuracy: 0.8508 - loss: 0.4161
Epoch 46/48
40/40 - 0s - 11ms/step - accuracy: 0.8550 - loss: 0.4055
Epoch 47/48
40/40 - 0s - 11ms/step - accuracy: 0.8528 - loss: 0.4061
Epoch 48/48
40/40 - 0s - 11ms/step - accuracy: 0.8561 - loss: 0.4026
10/10 - 0s - 20ms/step
Epoch 1/48
40/40 - 2s - 50ms/step - accuracy: 0.5844 - loss: 1.4276
Epoch 2/48
40/40 - 0s - 11ms/step - accuracy: 0.6701 - loss: 0.9779
Epoch 3/48
40/40 - 0s - 11ms/step - accuracy: 0.6915 - loss: 0.8990
Epoch 4/48
40/40 - 0s - 11ms/step - accuracy: 0.7071 - loss: 0.8316
Epoch 5/48
40/40 - 0s - 11ms/step - accuracy: 0.7182 - loss: 0.7928
Epoch 6/48
40/40 - 0s - 11ms/step - accuracy: 0.7345 - loss: 0.7524
Epoch 7/48
40/40 - 0s - 11ms/step - accuracy: 0.7405 - loss: 0.7310
Epoch 8/48
40/40 - 0s - 11ms/step - accuracy: 0.7511 - loss: 0.7014
Epoch 9/48
40/40 - 

40/40 - 0s - 11ms/step - accuracy: 0.7973 - loss: 0.5691
Epoch 18/48
40/40 - 1s - 13ms/step - accuracy: 0.7937 - loss: 0.5622
Epoch 19/48
40/40 - 1s - 14ms/step - accuracy: 0.8015 - loss: 0.5447
Epoch 20/48
40/40 - 1s - 17ms/step - accuracy: 0.8023 - loss: 0.5469
Epoch 21/48
40/40 - 1s - 27ms/step - accuracy: 0.8062 - loss: 0.5351
Epoch 22/48
40/40 - 0s - 11ms/step - accuracy: 0.8080 - loss: 0.5274
Epoch 23/48
40/40 - 0s - 11ms/step - accuracy: 0.8089 - loss: 0.5202
Epoch 24/48
40/40 - 0s - 11ms/step - accuracy: 0.8115 - loss: 0.5160
Epoch 25/48
40/40 - 0s - 11ms/step - accuracy: 0.8179 - loss: 0.5043
Epoch 26/48
40/40 - 0s - 11ms/step - accuracy: 0.8166 - loss: 0.5059
Epoch 27/48
40/40 - 0s - 11ms/step - accuracy: 0.8188 - loss: 0.4948
Epoch 28/48
40/40 - 0s - 11ms/step - accuracy: 0.8192 - loss: 0.4919
Epoch 29/48
40/40 - 0s - 11ms/step - accuracy: 0.8243 - loss: 0.4890
Epoch 30/48
40/40 - 0s - 11ms/step - accuracy: 0.8233 - loss: 0.4817
Epoch 31/48
40/40 - 0s - 11ms/step - accuracy:

Epoch 27/28
34/34 - 3s - 83ms/step - accuracy: 0.8763 - loss: 0.3575
Epoch 28/28
34/34 - 3s - 83ms/step - accuracy: 0.9008 - loss: 0.2816
9/9 - 0s - 39ms/step
Epoch 1/28
34/34 - 5s - 144ms/step - accuracy: 0.6130 - loss: 1.3195
Epoch 2/28
34/34 - 3s - 84ms/step - accuracy: 0.7175 - loss: 0.8365
Epoch 3/28
34/34 - 4s - 104ms/step - accuracy: 0.7510 - loss: 0.7398
Epoch 4/28
34/34 - 3s - 87ms/step - accuracy: 0.7714 - loss: 0.6748
Epoch 5/28
34/34 - 3s - 84ms/step - accuracy: 0.7779 - loss: 0.6407
Epoch 6/28
34/34 - 3s - 84ms/step - accuracy: 0.7899 - loss: 0.5956
Epoch 7/28
34/34 - 5s - 150ms/step - accuracy: 0.8031 - loss: 0.5616
Epoch 8/28
34/34 - 5s - 151ms/step - accuracy: 0.8068 - loss: 0.5498
Epoch 9/28
34/34 - 3s - 84ms/step - accuracy: 0.8168 - loss: 0.5198
Epoch 10/28
34/34 - 3s - 83ms/step - accuracy: 0.8198 - loss: 0.5110
Epoch 11/28
34/34 - 3s - 84ms/step - accuracy: 0.8324 - loss: 0.4711
Epoch 12/28
34/34 - 3s - 83ms/step - accuracy: 0.8362 - loss: 0.4649
Epoch 13/28
34/34 

Epoch 13/43
17/17 - 2s - 112ms/step - accuracy: 0.6821 - loss: 0.9115
Epoch 14/43
17/17 - 2s - 100ms/step - accuracy: 0.6883 - loss: 0.9065
Epoch 15/43
17/17 - 2s - 96ms/step - accuracy: 0.6902 - loss: 0.8889
Epoch 16/43
17/17 - 2s - 100ms/step - accuracy: 0.6947 - loss: 0.8865
Epoch 17/43
17/17 - 2s - 102ms/step - accuracy: 0.6943 - loss: 0.8786
Epoch 18/43
17/17 - 2s - 100ms/step - accuracy: 0.6972 - loss: 0.8736
Epoch 19/43
17/17 - 2s - 99ms/step - accuracy: 0.7049 - loss: 0.8634
Epoch 20/43
17/17 - 2s - 98ms/step - accuracy: 0.7020 - loss: 0.8638
Epoch 21/43
17/17 - 2s - 104ms/step - accuracy: 0.7011 - loss: 0.8626
Epoch 22/43
17/17 - 2s - 119ms/step - accuracy: 0.7029 - loss: 0.8546
Epoch 23/43
17/17 - 2s - 112ms/step - accuracy: 0.7091 - loss: 0.8528
Epoch 24/43
17/17 - 2s - 98ms/step - accuracy: 0.7101 - loss: 0.8495
Epoch 25/43
17/17 - 2s - 103ms/step - accuracy: 0.7072 - loss: 0.8448
Epoch 26/43
17/17 - 2s - 96ms/step - accuracy: 0.7081 - loss: 0.8394
Epoch 27/43
17/17 - 3s - 

Epoch 2/43
17/17 - 1s - 74ms/step - accuracy: 0.6528 - loss: 1.1238
Epoch 3/43
17/17 - 1s - 69ms/step - accuracy: 0.6545 - loss: 1.0751
Epoch 4/43
17/17 - 1s - 71ms/step - accuracy: 0.6593 - loss: 1.0462
Epoch 5/43
17/17 - 1s - 69ms/step - accuracy: 0.6631 - loss: 1.0254
Epoch 6/43
17/17 - 1s - 72ms/step - accuracy: 0.6670 - loss: 1.0036
Epoch 7/43
17/17 - 907s - 53s/step - accuracy: 0.6693 - loss: 0.9865
Epoch 8/43
17/17 - 2s - 144ms/step - accuracy: 0.6702 - loss: 0.9773
Epoch 9/43
17/17 - 1s - 87ms/step - accuracy: 0.6741 - loss: 0.9529
Epoch 10/43
17/17 - 1s - 79ms/step - accuracy: 0.6743 - loss: 0.9499
Epoch 11/43
17/17 - 1s - 61ms/step - accuracy: 0.6807 - loss: 0.9344
Epoch 12/43
17/17 - 1s - 55ms/step - accuracy: 0.6838 - loss: 0.9252
Epoch 13/43
17/17 - 1s - 53ms/step - accuracy: 0.6798 - loss: 0.9177
Epoch 14/43
17/17 - 1s - 54ms/step - accuracy: 0.6874 - loss: 0.9048
Epoch 15/43
17/17 - 1s - 52ms/step - accuracy: 0.6935 - loss: 0.8949
Epoch 16/43
17/17 - 1s - 53ms/step - acc

Epoch 26/47
30/30 - 2s - 65ms/step - accuracy: 0.2177 - loss: 2.4739
Epoch 27/47
30/30 - 1s - 38ms/step - accuracy: 0.2280 - loss: 2.4645
Epoch 28/47
30/30 - 1s - 38ms/step - accuracy: 0.2386 - loss: 2.4557
Epoch 29/47
30/30 - 1s - 38ms/step - accuracy: 0.2550 - loss: 2.4455
Epoch 30/47
30/30 - 1s - 38ms/step - accuracy: 0.2639 - loss: 2.4390
Epoch 31/47
30/30 - 1s - 38ms/step - accuracy: 0.2769 - loss: 2.4284
Epoch 32/47
30/30 - 2s - 56ms/step - accuracy: 0.2903 - loss: 2.4164
Epoch 33/47
30/30 - 2s - 52ms/step - accuracy: 0.2992 - loss: 2.4105
Epoch 34/47
30/30 - 2s - 52ms/step - accuracy: 0.3126 - loss: 2.3986
Epoch 35/47
30/30 - 2s - 52ms/step - accuracy: 0.3281 - loss: 2.3876
Epoch 36/47
30/30 - 2s - 57ms/step - accuracy: 0.3375 - loss: 2.3801
Epoch 37/47
30/30 - 2s - 64ms/step - accuracy: 0.3488 - loss: 2.3695
Epoch 38/47
30/30 - 2s - 63ms/step - accuracy: 0.3609 - loss: 2.3590
Epoch 39/47
30/30 - 2s - 67ms/step - accuracy: 0.3671 - loss: 2.3503
Epoch 40/47
30/30 - 2s - 66ms/step

30/30 - 2s - 79ms/step - accuracy: 0.2359 - loss: 2.5729
Epoch 4/47
30/30 - 2s - 71ms/step - accuracy: 0.2421 - loss: 2.5649
Epoch 5/47
30/30 - 1076s - 36s/step - accuracy: 0.2548 - loss: 2.5572
Epoch 6/47
30/30 - 2s - 72ms/step - accuracy: 0.2677 - loss: 2.5494
Epoch 7/47
30/30 - 2s - 59ms/step - accuracy: 0.2859 - loss: 2.5405
Epoch 8/47
30/30 - 2s - 71ms/step - accuracy: 0.2914 - loss: 2.5343
Epoch 9/47
30/30 - 1s - 44ms/step - accuracy: 0.3030 - loss: 2.5260
Epoch 10/47
30/30 - 1s - 45ms/step - accuracy: 0.3214 - loss: 2.5189
Epoch 11/47
30/30 - 1s - 44ms/step - accuracy: 0.3359 - loss: 2.5094
Epoch 12/47
30/30 - 4s - 131ms/step - accuracy: 0.3480 - loss: 2.5011
Epoch 13/47
30/30 - 3s - 90ms/step - accuracy: 0.3608 - loss: 2.4935
Epoch 14/47
30/30 - 2s - 67ms/step - accuracy: 0.3766 - loss: 2.4823
Epoch 15/47
30/30 - 2s - 62ms/step - accuracy: 0.3888 - loss: 2.4751
Epoch 16/47
30/30 - 2s - 77ms/step - accuracy: 0.4023 - loss: 2.4658
Epoch 17/47
30/30 - 2s - 53ms/step - accuracy: 0.

Epoch 7/21
26/26 - 1s - 25ms/step - accuracy: 0.7020 - loss: 0.8917
Epoch 8/21
26/26 - 1s - 26ms/step - accuracy: 0.7076 - loss: 0.8748
Epoch 9/21
26/26 - 1s - 24ms/step - accuracy: 0.7134 - loss: 0.8612
Epoch 10/21
26/26 - 1s - 22ms/step - accuracy: 0.7167 - loss: 0.8465
Epoch 11/21
26/26 - 1s - 23ms/step - accuracy: 0.7217 - loss: 0.8332
Epoch 12/21
26/26 - 1s - 28ms/step - accuracy: 0.7288 - loss: 0.8194
Epoch 13/21
26/26 - 1s - 29ms/step - accuracy: 0.7318 - loss: 0.8071
Epoch 14/21
26/26 - 1s - 27ms/step - accuracy: 0.7358 - loss: 0.7966
Epoch 15/21
26/26 - 1s - 29ms/step - accuracy: 0.7399 - loss: 0.7835
Epoch 16/21
26/26 - 1s - 29ms/step - accuracy: 0.7415 - loss: 0.7736
Epoch 17/21
26/26 - 1s - 26ms/step - accuracy: 0.7466 - loss: 0.7625
Epoch 18/21
26/26 - 1s - 23ms/step - accuracy: 0.7502 - loss: 0.7532
Epoch 19/21
26/26 - 1s - 24ms/step - accuracy: 0.7539 - loss: 0.7432
Epoch 20/21
26/26 - 1s - 24ms/step - accuracy: 0.7573 - loss: 0.7339
Epoch 21/21
26/26 - 1s - 25ms/step - 

Epoch 31/48
53/53 - 1s - 22ms/step - accuracy: 0.6440 - loss: 1.0773
Epoch 32/48
53/53 - 1s - 20ms/step - accuracy: 0.6440 - loss: 1.0732
Epoch 33/48
53/53 - 1s - 20ms/step - accuracy: 0.6440 - loss: 1.0701
Epoch 34/48
53/53 - 1s - 19ms/step - accuracy: 0.6440 - loss: 1.0668
Epoch 35/48
53/53 - 1s - 21ms/step - accuracy: 0.6440 - loss: 1.0638
Epoch 36/48
53/53 - 1s - 20ms/step - accuracy: 0.6440 - loss: 1.0609
Epoch 37/48
53/53 - 1s - 21ms/step - accuracy: 0.6440 - loss: 1.0587
Epoch 38/48
53/53 - 1s - 22ms/step - accuracy: 0.6440 - loss: 1.0566
Epoch 39/48
53/53 - 1s - 24ms/step - accuracy: 0.6440 - loss: 1.0543
Epoch 40/48
53/53 - 1s - 26ms/step - accuracy: 0.6440 - loss: 1.0525
Epoch 41/48
53/53 - 1s - 21ms/step - accuracy: 0.6440 - loss: 1.0504
Epoch 42/48
53/53 - 1s - 21ms/step - accuracy: 0.6440 - loss: 1.0486
Epoch 43/48
53/53 - 1s - 22ms/step - accuracy: 0.6440 - loss: 1.0474
Epoch 44/48
53/53 - 1s - 21ms/step - accuracy: 0.6440 - loss: 1.0454
Epoch 45/48
53/53 - 1s - 18ms/step

53/53 - 1s - 12ms/step - accuracy: 0.6439 - loss: 1.1615
Epoch 6/48
53/53 - 1s - 13ms/step - accuracy: 0.6439 - loss: 1.1506
Epoch 7/48
53/53 - 1s - 25ms/step - accuracy: 0.6439 - loss: 1.1419
Epoch 8/48
53/53 - 1s - 14ms/step - accuracy: 0.6439 - loss: 1.1343
Epoch 9/48
53/53 - 1s - 18ms/step - accuracy: 0.6439 - loss: 1.1278
Epoch 10/48
53/53 - 1s - 20ms/step - accuracy: 0.6439 - loss: 1.1218
Epoch 11/48
53/53 - 1s - 18ms/step - accuracy: 0.6439 - loss: 1.1166
Epoch 12/48
53/53 - 1s - 15ms/step - accuracy: 0.6439 - loss: 1.1116
Epoch 13/48
53/53 - 1s - 15ms/step - accuracy: 0.6439 - loss: 1.1067
Epoch 14/48
53/53 - 1s - 15ms/step - accuracy: 0.6439 - loss: 1.1024
Epoch 15/48
53/53 - 1s - 16ms/step - accuracy: 0.6439 - loss: 1.0985
Epoch 16/48
53/53 - 1s - 15ms/step - accuracy: 0.6439 - loss: 1.0947
Epoch 17/48
53/53 - 1s - 16ms/step - accuracy: 0.6439 - loss: 1.0912
Epoch 18/48
53/53 - 1s - 16ms/step - accuracy: 0.6439 - loss: 1.0880
Epoch 19/48
53/53 - 1s - 24ms/step - accuracy: 0.6

Epoch 20/27
16/16 - 3s - 168ms/step - accuracy: 0.6439 - loss: nan
Epoch 21/27
16/16 - 1s - 89ms/step - accuracy: 0.6439 - loss: nan
Epoch 22/27
16/16 - 1s - 90ms/step - accuracy: 0.6439 - loss: nan
Epoch 23/27
16/16 - 1s - 90ms/step - accuracy: 0.6439 - loss: nan
Epoch 24/27
16/16 - 1s - 89ms/step - accuracy: 0.6439 - loss: nan
Epoch 25/27
16/16 - 1s - 90ms/step - accuracy: 0.6439 - loss: nan
Epoch 26/27
16/16 - 1s - 91ms/step - accuracy: 0.6439 - loss: nan
Epoch 27/27
16/16 - 3s - 158ms/step - accuracy: 0.6439 - loss: nan
4/4 - 0s - 76ms/step
Epoch 1/27
16/16 - 4s - 254ms/step - accuracy: 0.6440 - loss: nan
Epoch 2/27
16/16 - 2s - 103ms/step - accuracy: 0.6440 - loss: nan
Epoch 3/27
16/16 - 2s - 155ms/step - accuracy: 0.6440 - loss: nan
Epoch 4/27
16/16 - 2s - 106ms/step - accuracy: 0.6440 - loss: nan
Epoch 5/27
16/16 - 1s - 93ms/step - accuracy: 0.6440 - loss: nan
Epoch 6/27
16/16 - 3s - 160ms/step - accuracy: 0.6440 - loss: nan
Epoch 7/27
16/16 - 1s - 92ms/step - accuracy: 0.6440 -

Epoch 19/36
30/30 - 1s - 18ms/step - accuracy: 0.6440 - loss: 1.1599
Epoch 20/36
30/30 - 1s - 22ms/step - accuracy: 0.6440 - loss: 1.1581
Epoch 21/36
30/30 - 1s - 18ms/step - accuracy: 0.6440 - loss: 1.1523
Epoch 22/36
30/30 - 1s - 18ms/step - accuracy: 0.6440 - loss: 1.1354
Epoch 23/36
30/30 - 1s - 21ms/step - accuracy: 0.6440 - loss: 1.1042
Epoch 24/36
30/30 - 1s - 18ms/step - accuracy: 0.6440 - loss: 1.0663
Epoch 25/36
30/30 - 1s - 18ms/step - accuracy: 0.6439 - loss: 1.0437
Epoch 26/36
30/30 - 1s - 23ms/step - accuracy: 0.6433 - loss: 1.0294
Epoch 27/36
30/30 - 1s - 26ms/step - accuracy: 0.6419 - loss: 1.0170
Epoch 28/36
30/30 - 1s - 20ms/step - accuracy: 0.6433 - loss: 1.0072
Epoch 29/36
30/30 - 1s - 18ms/step - accuracy: 0.6423 - loss: 0.9979
Epoch 30/36
30/30 - 1s - 18ms/step - accuracy: 0.6417 - loss: 0.9870
Epoch 31/36
30/30 - 1s - 26ms/step - accuracy: 0.6425 - loss: 0.9764
Epoch 32/36
30/30 - 1s - 17ms/step - accuracy: 0.6433 - loss: 0.9675
Epoch 33/36
30/30 - 1s - 21ms/step

Epoch 30/36
30/30 - 1s - 18ms/step - accuracy: 0.6415 - loss: 1.0177
Epoch 31/36
30/30 - 1s - 20ms/step - accuracy: 0.6390 - loss: 1.0121
Epoch 32/36
30/30 - 1s - 29ms/step - accuracy: 0.6427 - loss: 1.0061
Epoch 33/36
30/30 - 1s - 29ms/step - accuracy: 0.6371 - loss: 1.0034
Epoch 34/36
30/30 - 1s - 33ms/step - accuracy: 0.6426 - loss: 0.9973
Epoch 35/36
30/30 - 1s - 24ms/step - accuracy: 0.6426 - loss: 0.9932
Epoch 36/36
30/30 - 1s - 23ms/step - accuracy: 0.6423 - loss: 0.9888
8/8 - 0s - 30ms/step
| 12        | 0.6289    | 1.079     | 470.1     | 0.9429    | 0.09696   | 35.56     | 2.406     | 1.727     | 2.944     | 0.9628    | 32.66     | 0.4972    | 2.106     |
Epoch 1/22
60/60 - 3s - 54ms/step - accuracy: 0.6478 - loss: 1.1485
Epoch 2/22
60/60 - 2s - 27ms/step - accuracy: 0.7184 - loss: 0.8151
Epoch 3/22
60/60 - 1s - 23ms/step - accuracy: 0.7499 - loss: 0.7223
Epoch 4/22
60/60 - 1s - 23ms/step - accuracy: 0.7707 - loss: 0.6676
Epoch 5/22
60/60 - 3s - 43ms/step - accuracy: 0.7797 -

Epoch 20/22
60/60 - 2s - 33ms/step - accuracy: 0.8666 - loss: 0.3762
Epoch 21/22
60/60 - 2s - 37ms/step - accuracy: 0.8694 - loss: 0.3674
Epoch 22/22
60/60 - 2s - 27ms/step - accuracy: 0.8751 - loss: 0.3553
15/15 - 0s - 24ms/step
| 13        | 0.8241    | 2.564     | 229.5     | 0.6096    | 0.1508    | 21.54     | 1.557     | 2.817     | 1.479     | 0.1534    | 54.05     | 0.9857    | 1.694     |
Epoch 1/31
18/18 - 4s - 219ms/step - accuracy: 0.6248 - loss: 1.1834
Epoch 2/31
18/18 - 2s - 124ms/step - accuracy: 0.7011 - loss: 0.8983
Epoch 3/31
18/18 - 3s - 149ms/step - accuracy: 0.7257 - loss: 0.8065
Epoch 4/31
18/18 - 2s - 138ms/step - accuracy: 0.7418 - loss: 0.7551
Epoch 5/31
18/18 - 2s - 131ms/step - accuracy: 0.7552 - loss: 0.7127
Epoch 6/31
18/18 - 3s - 141ms/step - accuracy: 0.7388 - loss: 0.7302
Epoch 7/31
18/18 - 2s - 122ms/step - accuracy: 0.7563 - loss: 0.6950
Epoch 8/31
18/18 - 2s - 124ms/step - accuracy: 0.7773 - loss: 0.6473
Epoch 9/31
18/18 - 2s - 120ms/step - accuracy: 0

KeyboardInterrupt: 

In [24]:
optimum = nn_opt.max['params']
learning_rate = optimum['learning_rate']

activationL = ['relu', 'sigmoid', 'softplus', 'softsign', 'tanh', 'selu', 'elu', 'exponential', LeakyReLU, 'relu']
optimum['activation'] = activationL[round(optimum['activation'])]

optimum['batch_size'] = round(optimum['batch_size'])
optimum['epochs'] = round(optimum['epochs'])
optimum['layers1'] = round(optimum['layers1'])
optimum['layers2'] = round(optimum['layers2'])
optimum['neurons'] = round(optimum['neurons'])

optimizerL = ['Adam', 'SGD', 'RMSprop', 'Adadelta', 'Adagrad', 'Adamax', 'Nadam', 'Ftrl', 'Adam']
optimizerD = {
    'Adam': Adam(learning_rate=learning_rate),
    'SGD': SGD(learning_rate=learning_rate),
    'RMSprop': RMSprop(learning_rate=learning_rate),
    'Adadelta': Adadelta(learning_rate=learning_rate),
    'Adagrad': Adagrad(learning_rate=learning_rate),
    'Adamax': Adamax(learning_rate=learning_rate),
    'Nadam': Nadam(learning_rate=learning_rate),
    'Ftrl': Ftrl(learning_rate=learning_rate)
}
optimum['optimizer'] = optimizerD[optimizerL[round(optimum['optimizer'])]]
optimum

{'activation': 'softsign',
 'batch_size': 460,
 'dropout': 0.7296061783380641,
 'dropout_rate': 0.19126724140656393,
 'epochs': 47,
 'kernel': 1.9444298503238986,
 'layers1': 1,
 'layers2': 2,
 'learning_rate': 0.7631771981307285,
 'neurons': 61,
 'normalization': 0.770967179954561,
 'optimizer': <keras.src.optimizers.adadelta.Adadelta at 0x10bd87790>}

## 6. Running CNN with Optimized Search Parameters

In [25]:
# Set the model with optimized hyperparameters

epochs = 47
batch_size = 460

timesteps = len(X_train[0])
input_dim = len(X_train[0][0])
n_classes = 15

layers1 = 1
layers2 = 2
activation = 'softsign'
kernel = int(round(1.9444298503238986))  # Rounded kernel size for Conv1D
neurons = 61
normalization = 0.770967179954561
dropout = 0.7296061783380641
dropout_rate = 0.19126724140656393
optimizer = Adadelta(learning_rate=0.7631771981307285)  # Instantiate RMSprop with learning rate

model = Sequential()
model.add(Conv1D(neurons, kernel_size=kernel, activation=activation, input_shape=(timesteps, input_dim)))

if normalization > 0.5:
    model.add(BatchNormalization())

for i in range(layers1):
    model.add(Dense(neurons, activation=activation))

if dropout > 0.5:
    model.add(Dropout(dropout_rate))

for i in range(layers2):
    model.add(Dense(neurons, activation=activation))

model.add(MaxPooling1D())
model.add(Flatten())
model.add(Dense(n_classes, activation='softmax')) 

model.compile(loss='sparse_categorical_crossentropy', optimizer=optimizer, metrics=['accuracy'])

In [26]:
model.summary()

Model: "sequential_67"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ conv1d_67 (Conv1D)              │ (None, 14, 61)         │         1,159 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ batch_normalization_25          │ (None, 14, 61)         │           244 │
│ (BatchNormalization)            │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_345 (Dense)               │ (None, 14, 61)         │         3,782 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout_40 (Dropout)            │ (None, 14, 61)         │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_346 (Dense)               │ (None, 14, 61)         │         3,782 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_347 (Dense)               │ (None, 14, 61)         │         3,782 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling1d_67 (MaxPooling1D) │ (None, 7, 61)          │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ flatten_67 (Flatten)            │ (None, 427)            │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_348 (Dense)               │ (None, 15)             │         6,420 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 19,169 (74.88 KB)

 Trainable params: 19,047 (74.40 KB)

 Non-trainable params: 122 (488.00 B)

In [27]:
# Put the y_test set back into a one-hot configuration

y_train_one_hot = to_categorical(y_train, num_classes=15)

In [28]:
# Check shapes

print(f'X_train shape: {X_train.shape}')
print(f'y_train_one_hot shape: {y_train_one_hot.shape}')

X_train shape: (17212, 15, 9)
y_train_one_hot shape: (17212, 15)


In [29]:
# Compile the model with categorical_crossentropy

model.compile(loss='categorical_crossentropy', optimizer=optimizer, metrics=['accuracy'])

In [30]:
# Fit the model to the data

model.fit(X_train, y_train_one_hot, batch_size=batch_size, epochs=epochs, verbose=2)

Epoch 1/47
38/38 - 3s - 90ms/step - accuracy: 0.5807 - loss: 1.4373
Epoch 2/47
38/38 - 3s - 67ms/step - accuracy: 0.6932 - loss: 0.9095
Epoch 3/47
38/38 - 2s - 41ms/step - accuracy: 0.7144 - loss: 0.8258
Epoch 4/47
38/38 - 2s - 46ms/step - accuracy: 0.7406 - loss: 0.7634
Epoch 5/47
38/38 - 2s - 46ms/step - accuracy: 0.7538 - loss: 0.7168
Epoch 6/47
38/38 - 2s - 45ms/step - accuracy: 0.7707 - loss: 0.6781
Epoch 7/47
38/38 - 2s - 48ms/step - accuracy: 0.7804 - loss: 0.6427
Epoch 8/47
38/38 - 2s - 46ms/step - accuracy: 0.7877 - loss: 0.6143
Epoch 9/47
38/38 - 2s - 65ms/step - accuracy: 0.7996 - loss: 0.5851
Epoch 10/47
38/38 - 2s - 48ms/step - accuracy: 0.8074 - loss: 0.5578
Epoch 11/47
38/38 - 2s - 43ms/step - accuracy: 0.8150 - loss: 0.5315
Epoch 12/47
38/38 - 2s - 44ms/step - accuracy: 0.8194 - loss: 0.5166
Epoch 13/47
38/38 - 1s - 39ms/step - accuracy: 0.8259 - loss: 0.5005
Epoch 14/47
38/38 - 2s - 42ms/step - accuracy: 0.8322 - loss: 0.4816
Epoch 15/47
38/38 - 1s - 39ms/step - accura

## 7. Creating Confusion Matrix

In [31]:
# Define list of stations names

stations = {
0: 'BASEL',
1: 'BELGRADE',
2: 'BUDAPEST',
3: 'DEBILT',
4: 'DUSSELDORF',
5: 'HEATHROW',
6: 'KASSEL',
7: 'LJUBLJANA',
8: 'MAASTRICHT',
9: 'MADRID',
10: 'MUNCHENB',
11: 'OSLO',
12: 'SONNBLICK',
13: 'STOCKHOLM',
14: 'VALENTIA'
}

In [32]:
def confusion_matrix(y_true, y_pred, stations):
    # Check if y_true and y_pred are one-hot encoded or already class indices
    if y_true.ndim == 1:
        y_true_labels = y_true
    else:
        y_true_labels = np.argmax(y_true, axis=1)
    
    if y_pred.ndim == 1:
        y_pred_labels = y_pred
    else:
        y_pred_labels = np.argmax(y_pred, axis=1)
        
    # Map numeric labels to activity names
    y_true_series = pd.Series([stations[y] for y in y_true_labels])
    y_pred_series = pd.Series([stations[y] for y in y_pred_labels])
    
    return pd.crosstab(y_true_series, y_pred_series, rownames=['True'], colnames=['Pred'])

In [33]:
y_pred = model.predict(X_test)

180/180 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step


In [34]:
# Evaluate

print(confusion_matrix(y_test, y_pred, stations))

Pred        BASEL  BELGRADE  BUDAPEST  DEBILT  DUSSELDORF  HEATHROW  KASSEL  \
True                                                                          
BASEL        3517        93        11       4           3         3       1   
BELGRADE      105       987         0       0           0         0       0   
BUDAPEST       28        36       148       2           0         0       0   
DEBILT         14         6        27      35           0         0       0   
DUSSELDORF      5         2         6       5           5         5       0   
HEATHROW       10         2         4       3           2        60       0   
KASSEL          0         3         1       0           1         1       2   
LJUBLJANA      11         6         4       0           0         4       0   
MAASTRICHT      8         0         0       0           0         0       0   
MADRID         26        19        13       1           2        25       0   
MUNCHENB        5         1         0       0       